### ISSUE TITLE
Test failures (64 bit) with FLINT and ARB on Windows 7 

### ISSUE THREAD

**Issue by duhadler** [wbhart/flint2#128]
Dear FLINT and ARB Team!
I would like to report my experience with FLINT and ARB on Windows 7 (64 bit)
I am using the following system:
Environment: MSYS2 32 bit .
Compiler: TDM-GCC 4.9.2 (12 December 2014). Note: I have tested this compiler extensively; it compiles e.g.  Boost Math and Boost Multiprecision (MPFR and MPFI) without problems, and also works well with EIGEN (including the MPFR version).
GMP 5.1.3, configured as follows  
32 bit:
./configure --build=i686-w64-mingw32 --host=i686-w64-mingw32 --enable-fat --enable-cxx  --enable-static --disable-shared ABI=32  CFLAGS="-m32 -march=i686 -Wno-attributes" --prefix=/local/Win32 --exec-prefix=/local/Win32
64 bit:
./configure --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --enable-fat  --enable-static --disable-shared ABI=64  CFLAGS="-m64 -march=k8 -Wno-attributes" --prefix=/local/Win64 --exec-prefix=/local/Win64
MPFR 3.1.2, configured as follows  
32 bit:
./configure --build=i686-w64-mingw32 --host=i686-w64-mingw32  --enable-static --disable-shared   CFLAGS="-m32 -march=i686  -Wno-attributes" --with-gmp-include=/local/Win32/include --with-gmp-lib=/local/Win32/lib  --prefix=/local/Win32 --exec-prefix=/local/Win32
64 bit:
./configure --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32  --enable-static --disable-shared   CFLAGS="-m64 -march=k8  -Wno-attributes" --with-gmp-include=/local/Win64/include --with-gmp-lib=/local/Win64/lib  --prefix=/local/Win64 --exec-prefix=/local/Win64
FLINT 2.4.5, configured as follows  
32 bit:
./configure --build=x86-MINGW32  --enable-static --disable-shared ABI=32  CFLAGS="-m32 -march=i686 -Wno-attributes" --with-gmp=/local/Win32  --with-mpfr=/local/Win32  --prefix=/local/Win32
64 bit:
./configure --build=x86_64-MINGW64  --enable-static --disable-shared ABI=64  CFLAGS="-m64 -march=k8 -Wno-attributes" --with-gmp=/local/Win64  --with-mpfr=/local/Win64  --prefix=/local/Win64
ARB 2.5.0, configured as follows  
32 bit:
./configure --build=x86-MINGW32  --enable-static --disable-shared ABI=32  CFLAGS="-m32 -march=i686 -Wno-attributes" --with-gmp=/local/Win32  --with-mpfr=/local/Win32  --with-flint=/local/Win32 --prefix=/local/Win32
64 bit:
./configure --build=x86_64-MINGW64  --enable-static --disable-shared ABI=64  CFLAGS="-m64 -march=k8 -Wno-attributes" --with-gmp=/local/Win64  --with-mpfr=/local/Win64  --with-flint=/local/Win64  --prefix=/local/Win64

The outcome so far is the following:
GMP and MPFR compile without errors and warnings, and all tests pass (both 32 bit and 64 bit).
FLINT 32 bit compiles without errors and warnings, and all tests pass .
FLINT 64 bit compiles without errors and warnings, but 5 tests fail at run-time, all with a segmentation fault:
Fmpz\test\t-clog.c, fmpz\test\t-clog_ui.c, fmpz\test\t-flog.c, fmpz\test\t-flog_ui.c, fmpz_poly_factor\test\ t-factor_zassenhaus.c
With these tests removed, all other tests pass.

ARB 32 bit compiles without errors and warnings, and all but 2 tests pass. These tests are:
TEST1, Runtime error:
set_str....
This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
FAIL (valid input): inf
inf +/- 0
0 +/- 0
../Makefile.subdirs:84: recipe for target '../build/arb/test/t-set_str.exe_RUN'                                                                                                                                                                                                failed
make[1]: *** [../build/arb/test/t-set_str.exe_RUN] Error 3
make[1]: Leaving directory '/home/MP32/arb-2.5.0/arb'
Makefile:170: recipe for target 'check' failed
make: *** [check] Error 2

TEST2, Compile Time Error:
make[1]: Entering directory '/home/MP32/arb-2.5.0/acb_modular'
    CC   ../build/acb_modular/test/t-elliptic_p.exe
    CC   ../build/acb_modular/test/t-epsilon_arg.exe
I:\Data\mpFormulaTDM\ExternalTools\msys32\tmp\cc121Hff.o:t-epsilon_arg.c:(.text+0x490): undefined reference to `fmpq_dedekind_sum'
collect2.exe: error: ld returned 1 exit status
../Makefile.subdirs:78: recipe for target '../build/acb_modular/test/t-epsilon_arg.exe' failed
make[1]: *** [../build/acb_modular/test/t-epsilon_arg.exe] Error 1
make[1]: Leaving directory '/home/MP32/arb-2.5.0/acb_modular'
Makefile:170: recipe for target 'check' failed
make: *** [check] Error 2
When these 2 tests are removed, all other tests pass without error.

ARB 64 bit compiles without errors but with many  warnings. All tests seem to fail. It did not manage to try all tests, because run-time failure already occurs at the beginning, and it is very cumbersome to compile and run manually. This is a typical error message from the log:
DH@DH-PC MINGW32 /home/MP64/arb-2.5.0
$ make check
make[1]: Entering directory '/home/MP64/arb-2.5.0'
make[1]: Nothing to be done for 'static'.
make[1]: Leaving directory '/home/MP64/arb-2.5.0'
make[1]: Entering directory '/home/MP64/arb-2.5.0'
make[2]: Entering directory '/home/MP64/arb-2.5.0'
make[2]: Nothing to be done for 'static'.
make[2]: Leaving directory '/home/MP64/arb-2.5.0'
make[1]: Leaving directory '/home/MP64/arb-2.5.0'
make[1]: Entering directory '/home/MP64/arb-2.5.0/fmpr'
sqrt....
This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
FAIL

bits = 32
x = (1 * 2^0)

z = (1 * 2^0)

w = (-1 * 2^0)

../Makefile.subdirs:84: recipe for target '../build/fmpr/test/t-sqrt.exe_RUN' failed
make[1]: *** [../build/fmpr/test/t-sqrt.exe_RUN] Error 3
make[1]: Leaving directory '/home/MP64/arb-2.5.0/fmpr'
Makefile:170: recipe for target 'check' failed
make: *** [check] Error 2

Browsing through the source of various tests I noted that there is a lot of use of the types long and int. For 64 bit programs long is 64 bit on Linux but 32 bit on Windows, as I am sure you know, and this might in part explain these failures on the Windows platform.
The warnings during compilation seem to point in a similar direction:

DH@DH-PC MINGW32 /home/MP64/arb-2.5.0
$ bash arb_configure_Win64_Release.sh
Configuring...x86_64-MINGW64
Testing __builtin_popcountl...yes
Testing native popcount...no
Testing __thread...yes
ARB was successfully configured.

DH@DH-PC MINGW32 /home/MP64/arb-2.5.0
$ make
make[1]: Entering directory '/home/MP64/arb-2.5.0'
mkdir -p build/interfaces
make[2]: Entering directory '/home/MP64/arb-2.5.0/fmpr'
 
set_d.c: In function 'fmpr_set_d':
set_d.c:65:9: warning: left shift count >= width of type
         frac |= (1UL << 52);
         ^

make[2]: Leaving directory '/home/MP64/arb-2.5.0/fmpr'
make[2]: Entering directory '/home/MP64/arb-2.5.0/arf'
ceil.c: In function 'arf_ceil':
ceil.c:41:9: warning: right shift count >= width of type
         if (COEFF_IS_MPZ(exp) || exp <= 0)
         ^
set_d.c: In function 'arf_set_d':
set_d.c:66:9: warning: left shift count >= width of type
         frac |= (1UL << 52);
         ^
floor.c: In function 'arf_floor':
floor.c:41:9: warning: right shift count >= width of type
         if (COEFF_IS_MPZ(exp) || exp <= 0)
         ^
make[2]: Leaving directory '/home/MP64/arb-2.5.0/arf'
make[2]: Entering directory '/home/MP64/arb-2.5.0/mag'
set_fmpz_2exp_fmpz.c: In function 'mag_set_fmpz_2exp_fmpz':
set_fmpz_2exp_fmpz.c:40:37: warning: passing argument 1 of 'fmpz_abs_ubound_ui_2exp' from incompatible pointer type
         m = fmpz_abs_ubound_ui_2exp(&cexp, man, MAG_BITS);
                                     ^
In file included from I:/Data/mpFormulaTDM/ExternalTools/msys32/home/MP64/arb-2.5.0/mag.h:37:0,
                 from set_fmpz_2exp_fmpz.c:26:
I:/Data/mpFormulaTDM/ExternalTools/msys32/local/Win64/include/flint/fmpz.h:663:11: note: expected 'mp_limb_signed_t *' but argument is of type 'long int *'
 mp_limb_t fmpz_abs_ubound_ui_2exp(slong * exp, const fmpz_t x, int bits);
           ^
set_fmpz_2exp_fmpz.c: In function 'mag_set_fmpz_2exp_fmpz_lower':
set_fmpz_2exp_fmpz.c:58:37: warning: passing argument 1 of 'fmpz_abs_lbound_ui_2exp' from incompatible pointer type
         m = fmpz_abs_lbound_ui_2exp(&cexp, man, MAG_BITS);
                                     ^
In file included from I:/Data/mpFormulaTDM/ExternalTools/msys32/home/MP64/arb-2.5.0/mag.h:37:0,
                 from set_fmpz_2exp_fmpz.c:26:
I:/Data/mpFormulaTDM/ExternalTools/msys32/local/Win64/include/flint/fmpz.h:665:11: note: expected 'mp_limb_signed_t *' but argument is of type 'long int *'
 mp_limb_t fmpz_abs_lbound_ui_2exp(slong * exp, const fmpz_t x, int bits);
           ^

make[2]: Entering directory '/home/MP64/arb-2.5.0/arb_mat'
solve.c: In function 'arb_mat_solve':
solve.c:41:10: warning: assignment from incompatible pointer type
     perm = _perm_init(n);
          ^
solve.c:50:17: warning: passing argument 1 of '_perm_clear' from incompatible pointer type
     _perm_clear(perm);
                 ^
In file included from I:/Data/mpFormulaTDM/ExternalTools/msys32/home/MP64/arb-2.5.0/arb_mat.h:38:0,
                 from solve.c:26:
I:/Data/mpFormulaTDM/ExternalTools/msys32/local/Win64/include/flint/perm.h:63:24: note: expected 'mp_limb_signed_t *' but argument is of type 'long int *'
 static __inline__ void _perm_clear(slong * vec)
                        ^

make[2]: Entering directory '/home/MP64/arb-2.5.0/acb_mat'
solve.c: In function 'acb_mat_solve':
solve.c:41:10: warning: assignment from incompatible pointer type
     perm = _perm_init(n);
          ^
solve.c:50:17: warning: passing argument 1 of '_perm_clear' from incompatible pointer type
     _perm_clear(perm);
                 ^
In file included from I:/Data/mpFormulaTDM/ExternalTools/msys32/home/MP64/arb-2.5.0/arb_mat.h:38:0,
                 from I:/Data/mpFormulaTDM/ExternalTools/msys32/home/MP64/arb-2.5.0/acb_mat.h:39,
                 from solve.c:26:
I:/Data/mpFormulaTDM/ExternalTools/msys32/local/Win64/include/flint/perm.h:63:24: note: expected 'mp_limb_signed_t *' but argument is of type 'long int *'
 static __inline__ void _perm_clear(slong * vec)
                        ^

make[2]: Entering directory '/home/MP64/arb-2.5.0/fmprb'
    CC   ../build/fmprb/fmprb_mul.o
mul.c: In function '_rad_bound':
mul.c:60:34: warning: passing argument 1 of 'fmpz_abs_ubound_ui_2exp' from incompatible pointer type
     *m = fmpz_abs_ubound_ui_2exp(&e, fmpr_manref(t), FMPRB_RAD_PREC);
                                  ^
In file included from I:/Data/mpFormulaTDM/ExternalTools/msys32/home/MP64/arb-2.5.0/fmpr.h:34:0,
                 from I:/Data/mpFormulaTDM/ExternalTools/msys32/home/MP64/arb-2.5.0/fmprb.h:29,
                 from mul.c:26:
I:/Data/mpFormulaTDM/ExternalTools/msys32/local/Win64/include/flint/fmpz.h:663:11: note: expected 'mp_limb_signed_t *' but argument is of type 'long int *'
 mp_limb_t fmpz_abs_ubound_ui_2exp(slong * exp, const fmpz_t x, int bits);
           ^
    CC   ../build/fmprb/fmprb_contains_fmpr.o
make[2]: Leaving directory '/home/MP64/arb-2.5.0/fmprb'
make[2]: Entering directory '/home/MP64/arb-2.5.0/bernoulli'
    CC   ../build/bernoulli/bernoulli_bound_2exp_si.o
bound_2exp_si.c: In function 'bernoulli_bound_2exp_si':
bound_2exp_si.c:93:9: warning: left shift count >= width of type
         if (hi || n > (1UL << (FLINT_BITS - 6)))
         ^

touch: cannot touch ‘test/t-*.c’: No such file or directory
touch: cannot touch ‘hypgeom/test/t-*.c’: No such file or directory
    AR   libarb.a
make[1]: Leaving directory '/home/MP64/arb-2.5.0'

DH@DH-PC MINGW32 /home/MP64/arb-2.5.0
$

Any advice on the above issues would be much appreciated!
Best regards,

Dietrich Hadler

**Comment by wbhart** [comment]
Hi Dietrich,

Thanks for these reports. Can you let us know whether you were compiling
flint-2.4.5 or flint from our GitHub repository.

It's been a while since we've done a flint release, and the MinGW toolchain
has moved on since then. It's probably time we updated. I'm not sure when
this will be done, but obviously we always update to the latest MinGW when
making a major release.

Due to having so many other commitments, we are obviously behind.

Bill.

**Comment by DUHadler** [comment]
Hi Bill,

I was using flint-2.4.5. 

I took a look at the GitHub repository. Everything compiles with the MinGW tool chain, but there seem to be a number of (additional) test failures. For the code in the GitHub repository I have not compiled a complete list of issues regarding the MinGW tool chain yet. 

Would this be of interest?

Best regards,

Dietrich

**Comment by wbhart** [comment]
Hi Dietrich,

We've been using flint from the GitHub repository with MSYS2 on 64 bit
machines. This hasn't been causing us any issues.

You are welcome to post a list of issues, but we are probably going to
require msys2 in future because we are having far fewer issues with it.

That doesn't mean there aren't any issues with the current flint trunk.
Code gets changed there fairly frequently, so anything could be broken.
It's really only the official releases that we test with msys.

Bill.

**Comment by DUHadler** [comment]
Hi Bill,

I am using MSYS2 (triggered by a report that you posted some time ago :-)) and it is working very well (actually, msys crashes on my system when trying to compile FLINT with ARB). 

So it might with regard to FLINT come down to a compiler issue. Which version of mingw64 are you using with all tests in the current trunk passing for a 64 bit target? I have noticed huge differences between GCC 4.7 to 4.8 to 4.9 with a number of numerical programs. 

Any particular compiler switches to pay attention to? Maybe I try your working solution first before sending a long list of meaningless error reports...

Best regards,
Dietrich

**Comment by wbhart** [comment]
Hi Dietrich,

that is interesting. I don't recall which version of the compiler I was
using. I'm on leave atm so don't have access to that machine, so can't
check. Sorry.

I must confess, in the end we were making binaries using the OpenSUSE cross
compilation tools.

It looks like we need to revisit compilation with the latest MinGW and see
what is going wrong.

Bill.

**Comment by wbhart** [comment]
I've checked all of the tests for flog, flog_ui, clog, clog_ui and factor_zassenhaus with valgrind on linux. No problems there, so the problems with flint are at least Windows specific. I'll investigate further.

**Comment by wbhart** [comment]
All those tests pass for me with a static build of all the libraries. I'll investigate with dynamic builds.

**Comment by wbhart** [comment]
Also no problem with a shared build of all the libraries under recent msys2.

We did fix some issues in mpir-2.7.0 just before the release that caused problems with shared builds under msys2. They could have easily caused this problem.

I've also recently fixed many flint functions for Windows 64 (I don't claim to have fixed them all yet!!)

To make any further progress with the reported failures in flint, under Windows 64, we'd need:

* a confirmation the issues still exist under recent msys2
* some information about precisely where the error is occurring (e.g. place some printf traces in the code or run gdb)

I haven't looked at any of the Arb issues. Certainly Arb does not support Windows 64 yet. I'll leave it to the Arb author to check the 32 bit issues.
